// @databricks/sql/dist/telemetry/types.js
"use strict";
/**
 * Copyright (c) 2025 Databricks Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TELEMETRY_CONFIG = exports.TelemetryEventType = void 0;
/**
 * Event types emitted by the telemetry system
 */
var TelemetryEventType;
(function (TelemetryEventType) {
    TelemetryEventType["CONNECTION_OPEN"] = "connection.open";
    TelemetryEventType["STATEMENT_START"] = "statement.start";
    TelemetryEventType["STATEMENT_COMPLETE"] = "statement.complete";
    TelemetryEventType["CLOUDFETCH_CHUNK"] = "cloudfetch.chunk";
    TelemetryEventType["ERROR"] = "error";
})(TelemetryEventType = exports.TelemetryEventType || (exports.TelemetryEventType = {}));
/**
 * Default telemetry configuration values
 */
exports.DEFAULT_TELEMETRY_CONFIG = {
    enabled: false,
    batchSize: 100,
    flushIntervalMs: 5000,
    maxRetries: 3,
    authenticatedExport: true,
    circuitBreakerThreshold: 5,
    circuitBreakerTimeout: 60000, // 1 minute
};
