// @databricks/sql/dist/connection/auth/tokenProvider/ExternalTokenProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Token_1 = __importDefault(require("./Token"));
/**
 * A token provider that delegates token retrieval to an external callback function.
 * Useful for integrating with secret managers, vaults, or other token sources.
 */
class ExternalTokenProvider {
    /**
     * Creates a new ExternalTokenProvider.
     * @param getToken - Callback function that returns the access token string
     * @param options - Optional configuration
     * @param options.parseJWT - If true, attempt to extract expiration from JWT payload (default: true)
     * @param options.name - Custom name for this provider (default: "ExternalTokenProvider")
     */
    constructor(getToken, options) {
        var _a, _b;
        this.getTokenCallback = getToken;
        this.parseJWT = (_a = options === null || options === void 0 ? void 0 : options.parseJWT) !== null && _a !== void 0 ? _a : true;
        this.providerName = (_b = options === null || options === void 0 ? void 0 : options.name) !== null && _b !== void 0 ? _b : 'ExternalTokenProvider';
    }
    async getToken() {
        const accessToken = await this.getTokenCallback();
        if (this.parseJWT) {
            return Token_1.default.fromJWT(accessToken);
        }
        return new Token_1.default(accessToken);
    }
    getName() {
        return this.providerName;
    }
}
exports.default = ExternalTokenProvider;
