// @databricks/sql/dist/connection/auth/tokenProvider/FederationProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_fetch_1 = __importDefault(require("node-fetch"));
const Token_1 = __importDefault(require("./Token"));
const utils_1 = require("./utils");
/**
 * Token exchange endpoint path for Databricks OIDC.
 */
const TOKEN_EXCHANGE_ENDPOINT = '/oidc/v1/token';
/**
 * Grant type for RFC 8693 token exchange.
 */
const TOKEN_EXCHANGE_GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:token-exchange';
/**
 * Subject token type for JWT tokens.
 */
const SUBJECT_TOKEN_TYPE = 'urn:ietf:params:oauth:token-type:jwt';
/**
 * Default scope for SQL operations.
 */
const DEFAULT_SCOPE = 'sql';
/**
 * Timeout for token exchange requests in milliseconds.
 */
const REQUEST_TIMEOUT_MS = 30000;
/**
 * Maximum number of retry attempts for transient errors.
 */
const MAX_RETRY_ATTEMPTS = 3;
/**
 * Base delay in milliseconds for exponential backoff.
 */
const RETRY_BASE_DELAY_MS = 1000;
/**
 * HTTP status codes that are considered retryable.
 */
const RETRYABLE_STATUS_CODES = new Set([429, 500, 502, 503, 504]);
/**
 * Error class for token exchange failures that includes the HTTP status code.
 */
class TokenExchangeError extends Error {
    constructor(message, statusCode) {
        super(message);
        this.name = 'TokenExchangeError';
        this.statusCode = statusCode;
    }
}
/**
 * A token provider that wraps another provider with automatic token federation.
 * When the base provider returns a token from a different issuer, this provider
 * exchanges it for a Databricks-compatible token using RFC 8693.
 */
class FederationProvider {
    /**
     * Creates a new FederationProvider.
     * @param baseProvider - The underlying token provider
     * @param databricksHost - The Databricks workspace host URL
     * @param options - Optional configuration
     * @param options.clientId - Client ID for M2M/service principal federation
     * @param options.returnOriginalTokenOnFailure - Return original token if exchange fails (default: true)
     */
    constructor(baseProvider, databricksHost, options) {
        var _a;
        this.baseProvider = baseProvider;
        this.databricksHost = databricksHost;
        this.clientId = options === null || options === void 0 ? void 0 : options.clientId;
        this.returnOriginalTokenOnFailure = (_a = options === null || options === void 0 ? void 0 : options.returnOriginalTokenOnFailure) !== null && _a !== void 0 ? _a : true;
    }
    async getToken() {
        const token = await this.baseProvider.getToken();
        // Check if token needs exchange
        if (!this.needsTokenExchange(token)) {
            return token;
        }
        // Attempt token exchange
        try {
            return await this.exchangeToken(token);
        }
        catch (error) {
            if (this.returnOriginalTokenOnFailure) {
                // Fall back to original token
                return token;
            }
            throw error;
        }
    }
    getName() {
        return `federated[${this.baseProvider.getName()}]`;
    }
    /**
     * Determines if the token needs to be exchanged.
     * @param token - The token to check
     * @returns true if the token should be exchanged
     */
    needsTokenExchange(token) {
        const issuer = (0, utils_1.getJWTIssuer)(token.accessToken);
        // If we can't extract the issuer, don't exchange (might not be a JWT)
        if (!issuer) {
            return false;
        }
        // If the issuer is the same as Databricks host, no exchange needed
        if ((0, utils_1.isSameHost)(issuer, this.databricksHost)) {
            return false;
        }
        return true;
    }
    /**
     * Exchanges the token for a Databricks-compatible token using RFC 8693.
     * Includes retry logic for transient errors with exponential backoff.
     * @param token - The token to exchange
     * @returns The exchanged token
     */
    async exchangeToken(token) {
        return this.exchangeTokenWithRetry(token, 0);
    }
    /**
     * Attempts a single token exchange request.
     * @returns The exchanged token
     */
    async attemptTokenExchange(body) {
        var _a;
        const url = this.buildExchangeUrl();
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
        try {
            const response = await (0, node_fetch_1.default)(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded',
                },
                body,
                signal: controller.signal,
            });
            if (!response.ok) {
                const errorText = await response.text();
                const error = new TokenExchangeError(`Token exchange failed: ${response.status} ${response.statusText} - ${errorText}`, response.status);
                throw error;
            }
            const data = (await response.json());
            if (!data.access_token) {
                throw new Error('Token exchange response missing access_token');
            }
            // Calculate expiration from expires_in
            let expiresAt;
            if (typeof data.expires_in === 'number') {
                expiresAt = new Date(Date.now() + data.expires_in * 1000);
            }
            return new Token_1.default(data.access_token, {
                tokenType: (_a = data.token_type) !== null && _a !== void 0 ? _a : 'Bearer',
                expiresAt,
            });
        }
        finally {
            clearTimeout(timeoutId);
        }
    }
    /**
     * Recursively attempts token exchange with exponential backoff.
     */
    async exchangeTokenWithRetry(token, attempt) {
        const params = new URLSearchParams({
            grant_type: TOKEN_EXCHANGE_GRANT_TYPE,
            subject_token_type: SUBJECT_TOKEN_TYPE,
            subject_token: token.accessToken,
            scope: DEFAULT_SCOPE,
        });
        if (this.clientId) {
            params.append('client_id', this.clientId);
        }
        try {
            return await this.attemptTokenExchange(params.toString());
        }
        catch (error) {
            const canRetry = attempt < MAX_RETRY_ATTEMPTS && this.isRetryableError(error);
            if (!canRetry) {
                throw error;
            }
            // Exponential backoff: 1s, 2s, 4s
            const delay = RETRY_BASE_DELAY_MS * 2 ** attempt;
            await new Promise((resolve) => {
                setTimeout(resolve, delay);
            });
            return this.exchangeTokenWithRetry(token, attempt + 1);
        }
    }
    /**
     * Determines if an error is retryable (transient HTTP errors, network errors, timeouts).
     */
    isRetryableError(error) {
        if (error instanceof TokenExchangeError) {
            return RETRYABLE_STATUS_CODES.has(error.statusCode);
        }
        if (error instanceof Error) {
            return error.name === 'AbortError' || error.name === 'FetchError';
        }
        return false;
    }
    /**
     * Builds the token exchange URL.
     */
    buildExchangeUrl() {
        let host = this.databricksHost;
        // Ensure host has a protocol
        if (!host.includes('://')) {
            host = `https://${host}`;
        }
        // Remove trailing slash
        if (host.endsWith('/')) {
            host = host.slice(0, -1);
        }
        return `${host}${TOKEN_EXCHANGE_ENDPOINT}`;
    }
}
exports.default = FederationProvider;
