// @databricks/sql/dist/connection/auth/tokenProvider/StaticTokenProvider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Token_1 = __importDefault(require("./Token"));
/**
 * A token provider that returns a static token.
 * Useful for testing or when the token is obtained through external means.
 */
class StaticTokenProvider {
    /**
     * Creates a new StaticTokenProvider.
     * @param accessToken - The access token string
     * @param options - Optional token configuration (tokenType, expiresAt, refreshToken, scopes)
     */
    constructor(accessToken, options) {
        this.token = new Token_1.default(accessToken, options);
    }
    /**
     * Creates a StaticTokenProvider from a JWT string.
     * The expiration time will be extracted from the JWT payload.
     * @param jwt - The JWT token string
     * @param options - Optional token configuration
     */
    static fromJWT(jwt, options) {
        const token = Token_1.default.fromJWT(jwt, options);
        return new StaticTokenProvider(token.accessToken, {
            tokenType: token.tokenType,
            expiresAt: token.expiresAt,
            refreshToken: token.refreshToken,
            scopes: token.scopes,
        });
    }
    async getToken() {
        return this.token;
    }
    getName() {
        return 'StaticTokenProvider';
    }
}
exports.default = StaticTokenProvider;
