// @databricks/sql/dist/connection/auth/tokenProvider/Token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Safety buffer in seconds to consider a token expired before its actual expiration time.
 * This prevents using tokens that are about to expire during in-flight requests.
 */
const EXPIRATION_BUFFER_SECONDS = 30;
/**
 * Represents an access token with optional metadata and lifecycle management.
 */
class Token {
    constructor(accessToken, options) {
        var _a;
        this._accessToken = accessToken;
        this._tokenType = (_a = options === null || options === void 0 ? void 0 : options.tokenType) !== null && _a !== void 0 ? _a : 'Bearer';
        this._expiresAt = options === null || options === void 0 ? void 0 : options.expiresAt;
        this._refreshToken = options === null || options === void 0 ? void 0 : options.refreshToken;
        this._scopes = options === null || options === void 0 ? void 0 : options.scopes;
    }
    /**
     * The access token string.
     */
    get accessToken() {
        return this._accessToken;
    }
    /**
     * The token type (e.g., "Bearer").
     */
    get tokenType() {
        return this._tokenType;
    }
    /**
     * The expiration time of the token, if known.
     */
    get expiresAt() {
        return this._expiresAt;
    }
    /**
     * The refresh token, if available.
     */
    get refreshToken() {
        return this._refreshToken;
    }
    /**
     * The scopes associated with this token.
     */
    get scopes() {
        return this._scopes;
    }
    /**
     * Checks if the token has expired, including a safety buffer.
     * Returns false if expiration time is unknown.
     */
    isExpired() {
        if (!this._expiresAt) {
            return false;
        }
        const now = new Date();
        const bufferMs = EXPIRATION_BUFFER_SECONDS * 1000;
        return this._expiresAt.getTime() - bufferMs <= now.getTime();
    }
    /**
     * Sets the Authorization header on the provided headers object.
     * @param headers - The headers object to modify
     * @returns The modified headers object with Authorization set
     */
    setAuthHeader(headers) {
        return {
            ...headers,
            Authorization: `${this._tokenType} ${this._accessToken}`,
        };
    }
    /**
     * Creates a Token from a JWT string, extracting the expiration time from the payload.
     * If the JWT cannot be decoded, the token is created without expiration info.
     * The server will validate the token anyway, so decoding failures are handled gracefully.
     * @param jwt - The JWT token string
     * @param options - Additional token options (tokenType, refreshToken, scopes).
     *                  Note: expiresAt is not accepted here as it is extracted from the JWT payload.
     * @returns A new Token instance with expiration extracted from the JWT (if available)
     */
    static fromJWT(jwt, options) {
        let expiresAt;
        try {
            const parts = jwt.split('.');
            if (parts.length >= 2) {
                const payload = Buffer.from(parts[1], 'base64').toString('utf8');
                const decoded = JSON.parse(payload);
                if (typeof decoded.exp === 'number') {
                    expiresAt = new Date(decoded.exp * 1000);
                }
            }
        }
        catch (_a) {
            // If we can't decode the JWT, we'll proceed without expiration info
            // The server will validate the token anyway
        }
        return new Token(jwt, {
            tokenType: options === null || options === void 0 ? void 0 : options.tokenType,
            expiresAt,
            refreshToken: options === null || options === void 0 ? void 0 : options.refreshToken,
            scopes: options === null || options === void 0 ? void 0 : options.scopes,
        });
    }
    /**
     * Converts the token to a plain object for serialization.
     */
    toJSON() {
        var _a;
        return {
            accessToken: this._accessToken,
            tokenType: this._tokenType,
            expiresAt: (_a = this._expiresAt) === null || _a === void 0 ? void 0 : _a.toISOString(),
            refreshToken: this._refreshToken,
            scopes: this._scopes,
        };
    }
}
exports.default = Token;
